src/types/mod.rs

```diff
@@ -232,9 +288,57 @@ impl Decodable for Chat {
 
 impl Encodable for Chat {
     fn encode<E: Encoder>(&self, e: &mut E) -> Result<(), E::Error> {
+
         match self {
-            &Chat::User(ref u) => u.encode(e),
-            &Chat::Group(ref g) => g.encode(e),
+            &Chat::Private { ref id, ref first_name, ref last_name, ref username } => {
+                e.emit_struct("Chat", 5, |e| {
+                    try!(e.emit_struct_field("id", 0, |e| {
+                        id.encode(e)
+                    }));
+                    try!(e.emit_struct_field("type", 1, |e| {
+                        "private".encode(e)
+                    }));
+                    try!(e.emit_struct_field("first_name", 2, |e| {
+                        first_name.encode(e)
+                    }));
+                    try!(e.emit_struct_field("last_name", 3, |e| {
+                        last_name.encode(e)
+                    }));
+                    try!(e.emit_struct_field("type", 4, |e| {
+                        username.encode(e)
+                    }));
+                    Ok(())
+                })
+            },
+            &Chat::Group { ref id, ref title, ref is_supergroup} => {
+                e.emit_struct("Chat", 3, |e| {
+                    try!(e.emit_struct_field("id", 0, |e| {
+                        id.encode(e)
+                    }));
+                    try!(e.emit_struct_field("type", 1, |e| {
+                        let typ = if *is_supergroup { "supergroup" } else { "group" };
+                        typ.encode(e)
+                    }));
+                    try!(e.emit_struct_field("title", 2, |e| {
+                        title.encode(e)
+                    }));
+                    Ok(())
+                })
+            },
+            &Chat::Channel { ref id, ref title, ref name} => {
+                e.emit_struct("Channel", 3, |e| {
+                    try!(e.emit_struct_field("id", 0, |e| {
+                        id.encode(e)
+                    }));
+                    try!(e.emit_struct_field("title", 1, |e| {
+                        title.encode(e)
+                    }));
+                    try!(e.emit_struct_field("username", 2, |e| {
+                        name.encode(e)
+                    }));
```

Comment:
I think that you can just write `Encodable::encode` instead of all those `|e| { name.encode(e) }` closures. Not entirely sure though. If you do it you can also `use Encodeable::encode;` at the top of this function to shrink the code even further to: `try!(e.emit_struct_field("username", 2, encode))`.